test/test_writer.py

```diff
@@ -0,0 +1,207 @@
+    data['cat'] = data.hello.astype('category')
+
+    fname = os.path.join(tempdir, 'test.parquet')
+    write(fname, data, file_scheme=scheme, partitions=partitions,
+          compression=comp)
+
+    df = sql.read.parquet(fname)
+    ddf = df.toPandas()
+    for col in data:
+        assert (ddf[col] == data[col])[~ddf[col].isnull()].all()
+
+
+def test_roundtrip_s3():
+    s3fs = pytest.importorskip('s3fs')
+    s3 = s3fs.S3FileSystem()
+    data = pd.DataFrame({'i32': np.arange(1000, dtype=np.int32),
+                         'i64': np.arange(1000, dtype=np.int64),
+                         'f': np.arange(1000, dtype=np.float64),
+                         'bhello': np.random.choice([b'hello', b'you',
+                            b'people'], size=1000).astype("O")})
+    data['hello'] = data.bhello.str.decode('utf8')
+    data['bcat'] = data.bhello.astype('category')
+    data.loc[100, 'f'] = np.nan
+    data['cat'] = data.hello.astype('category')
+    noop = lambda x: True
+    myopen = lambda f: s3.open(f, 'wb')
+    write('MDtemp/temp_parq', data, file_scheme='hive', partitions=[0, 500],
+          open_with=myopen, mkdirs=noop)
+    myopen = s3.open
+    pf = ParquetFile('MDtemp/temp_parq', open_with=myopen)
+    df = pf.to_pandas(usecats=['cat', 'bcat'])
+    for col in data:
+        assert (df[col] == data[col])[~df[col].isnull()].all()
+
+    # for line in open('/Users/mdurant/.aws/credentials'):
+    #     if line.startswith('aws_access'):
+    #         key = line.split('=')[1][:-1]
+    #     if line.startswith('aws_secret'):
+    #         secret = line.split('=')[1][:-1]
+    #
+    # df = sql.read.parquet('s3a://{}:{}@MDtemp/temp_parq'.format(key, secret))
+    # ddf = df.toPandas()
+    # for col in data:
+    #     assert (ddf[col] == data[col])[~ddf[col].isnull()].all()
+
+
+@pytest.mark.parametrize('scheme,partitions,comp',
+                         product(('simple', 'hive'),
+                                 ([0], [0, 500]),
+                                 (None, 'GZIP', 'SNAPPY')))
+def test_roundtrip(tempdir, scheme, partitions, comp):
+    data = pd.DataFrame({'i32': np.arange(1000, dtype=np.int32),
+                         'i64': np.arange(1000, dtype=np.int64),
+                         'f': np.arange(1000, dtype=np.float64),
+                         'bhello': np.random.choice([b'hello', b'you',
+                            b'people'], size=1000).astype("O")})
+    data['a'] = np.array([b'a', b'b', b'c', b'd', b'e']*200, dtype="S1")
+    data['aa'] = data['a'].map(lambda x: 2*x).astype("S2")
+    data['hello'] = data.bhello.str.decode('utf8')
+    data['bcat'] = data.bhello.astype('category')
+    data['cat'] = data.hello.astype('category')
+    fname = os.path.join(tempdir, 'test.parquet')
+    write(fname, data, file_scheme=scheme, partitions=partitions,
+          compression=comp)
+
+    r = ParquetFile(fname)
+
+    df = r.to_pandas()
+    for col in r.columns:
+        assert (df[col] == data[col]).all()
```

Comment:
@jreback can you recommend a Pandas testing solution that would generate Series of all valid Pandas dtypes?  It would be nice to have something we could use in `pytest.mark.parametrize` to exhaustively test the kinds of pandas that that we might expect users to produce.  We might have to xfail a few to start, but it'd be good to have something comprehensive to work towards.